lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1550,6 +1557,86 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
   KPE->resolveComponents(TC.Context, components);
 }
 
+Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
+  // If constructor call is followed by `BOE` or `FVE` it means that such
+  // call is supposed to produce optional and literal constructors aren't
+  // failable so let's not attempt this optimization.
+  if (!ExprStack.empty()) {
+    auto *parent = ExprStack.back();
+    if (isa<BindOptionalExpr>(parent) || isa<ForceValueExpr>(parent))
+      return nullptr;
```

Comment:
If we don't to break source at all then yes, I think it's necessary since in this situation we can deduce that failable initializer is expected.

Reply:
In that case we should have a test that shows the difference in behavior we would get if this code isn't present, so that if someone does some refactoring or bug fixing in the future we ensure that we're still guarding against the behavioral change you envision would happen without this code.

Reply:
Sure, it's easy to add, I'll do that once I'm done with tests (`SwiftPM` would fail right now if I remove this check).